README.md

```diff
@@ -1,7 +1,117 @@
+[![Python3.7](https://img.shields.io/badge/python-3.7-red.svg)](https://www.python.org/)
+[![Python3.8](https://img.shields.io/badge/python-3.8-red.svg)](https://www.python.org/)
+[![GPL license](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/old-licenses/gpl-3.0.html)
+
+Cartesian genetic programming (CGP) in Python.
+
+This library implements Cartesian genetic programming in pure Python. It implements Python data structures to represent computational graphs in the Cartesian Genetic programming and provides one evolutionary algorithm,  (mu + lambda) evolution strategies, to evolve a population of computational graphs given an objective function.
+
+![CGP Sketch](cgp-sketch.png)
+
+Basic usage
+===========
+
+A complete example of CGP applied to a regression problem can be found in `examples/example_evo_regression.py`.
+
+The basic steps to be taken are:
+
+1. **Define an objective function**. 
+
+   The objective needs to take an individual as input variable and update the `fitness` member of the individual.
+```
+def objective(individual):
+    """Objective function of the regression task.
+
+    Parameters
+    ----------
+    individual : gp.CGPIndividual
+        Individual of the Cartesian Genetic Programming Framework.
+
+    Returns
+    -------
+    gp.CGPIndividual
+        Modified individual with updated fitness value.
+    """
+	# Compute the fitness value
+	fitness = ...
+	individual.fitness = fitness
+	return individual
+```
+2. Define parameters for the genome, the population and the evolutionary algorithm
+```
+params = {
+   'seed': 8188212,
+   'n_threads': 1,
+   'max_generations': 1000,
+   'min_fitness': 0.,
+
+   'population_params': {
+        'n_parents': 10,
+        'mutation_rate': 0.5,
+    },
+
+   'ea_params': {
+       'n_offsprings': 10,
+       'n_breeding': 10,
+       'tournament_size': 1},
+
+   'genome_params': {
+       'n_inputs': 2,
+       'n_outputs': 1,
+       'n_columns': 10,
+       'n_rows': 5,
+       'levels_back': 2,
+       'primitives': [gp.CGPAdd, gp.CGPSub, gp.CGPMul, gp.CGPDiv, gp.CGPConstantFloat]},
+    }
+```
+3. Initialize the `CGPPopulation` and the evolutionary algorithm instance:
+```
+pop = gp.CGPPopulation(**params['population_params'],
+                          seed=params['seed'], genome_params=params['genome_params'])
+ea = gp.ea.MuPlusLambda(**params['ea_params'])
+```
+4. Define a callback function to record information about the evolution of the population:
+```
+def record_history(pop, history):
+    keys = ['fitness']
+    for key in keys:
+        if key not in history:
+            history[key] = []
+    history['fitness'].append(pop.fitness_parents())
+```
+5. Use the `evolve` function that ties everything together and executes the evolution:
+```
+history = gp.evolve(pop, obj, ea, params['max_generations'],
+                       params['min_fitness'], record_history=record_history, print_progress=True)
+```
+
+
+
+Data structures
```

Comment:
do we need such a detailed discussion of the relevant classes here?

Reply:
Why not? I think it is useful for someone who likes to understand this library.

Reply:
my main concern is maintenance: as the library develops this might become outdated and we need to keep track of it manually. for such "low level" documentation maybe something automated like sphinx is the better option?